addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,487 @@
+ * <li>Execute Niko Home Control commands.
+ * <li>Listen to events from Niko Home Control.
+ * </ul>
+ *
+ * Only switch, dimmer and rollershutter actions are currently implemented.
+ *
+ * A class instance is instantiated from the {@link NikoHomeControlBridgeHandler} class initialization.
+ *
+ * @author Mark Herwege
+ */
+public final class NikoHomeControlCommunication {
+
+    private Logger logger = LoggerFactory.getLogger(NikoHomeControlCommunication.class);
+
+    private boolean fixedIp;
+    private InetAddress nhcAddress;
+    private int nhcPort;
+    private InetAddress broadcastAddr;
+
+    private Socket nhcSocket;
+    private PrintWriter nhcOut;
+    private BufferedReader nhcIn;
+
+    private boolean listenerStopped;
+    private boolean nhcEventsRunning;
+
+    // We keep only 2 gson adapters used to serialize and deserialize all messages sent and received
+    private Gson gsonOut = new Gson();
+    private Gson gsonIn;
+
+    private final NhcSystemInfo systemInfo = new NhcSystemInfo();
+    private final Map<Integer, NhcLocation> locations = new HashMap<>();
+    private final Map<Integer, NhcAction> actions = new HashMap<>();
+
+    private NikoHomeControlBridgeHandler bridgeCallBack;
+
+    /**
+     * Constructor for Niko Home Control communication object, manages communication with
+     * Niko Home Control IP-interface.
+     *
+     * @param addr Can be null or omitted, will attempt to discover IP address if omitted.
+     *
+     * @throws IOException when Niko Home Control IP-interface cannot be found
+     */
+    public NikoHomeControlCommunication(InetAddress addr, InetAddress broadcastAddr) throws IOException {
+
+        this.broadcastAddr = broadcastAddr;
+
+        if (addr == null) {
+            NikoHomeControlDiscover nhcDiscover = new NikoHomeControlDiscover(broadcastAddr);
+            this.nhcAddress = nhcDiscover.getAddr();
+            this.fixedIp = false;
+        } else {
+            this.nhcAddress = addr;
+            this.fixedIp = true;
+        }
+
+        // When we set up this object, we also want to get the proper gson adapter set up once
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(NhcMessageBase.class, new NikoHomeControlMessageDeserializer());
+        this.gsonIn = gsonBuilder.create();
+    }
+
+    public NikoHomeControlCommunication() throws IOException {
+
+        this(null, null);
+    }
+
+    /**
+     * Start communication with Niko Home Control IP-interface, run through initialization and start thread listening
+     * to all messages coming from Niko Home Control.
+     *
+     */
+    public void startCommunication(int port) {
+
+        try {
+            for (int i = 1; nhcEventsRunning && (i <= 5); i++) {
+                // the events listener thread did not finish yet, so wait max 5000ms before restarting
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException e) {
```

Comment:
Just move this catch above the other catch or use a multi catch, better not throw to catch in the same method